#### sklearn.model_selection.KFold

* class sklearn.model_selection.KFold(n_splits=5, *, shuffle=False, random_state=None)
n_splitsint, default=5
Number of folds. Must be at least 2

shufflebool, default=False 랜덤으로 섞음
Whether to shuffle the data before splitting into batches. Note that the samples within each split will not be shuffled.

random_stateint, RandomState instance or None, default=None
When shuffle is True, random_state affects the ordering of the indices, which controls the randomness of each fold. Otherwise, this parameter has no effect. Pass an int for reproducible output across multiple function calls. See Glossary.

In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

In [47]:
from sklearn.datasets import load_iris

In [11]:
iris=load_iris()
features=iris.data
label=iris.target
dt_clf = DecisionTreeClassifier(random_state=156)

In [32]:
kfold = KFold(n_splits=5)
print(kfold.split(features))

<generator object _BaseKFold.split at 0x0000028D6DC9FE20>


In [33]:
cv_accuracy = []
print('붓꽃 데이터 세트 크기', features.shape[0])

붓꽃 데이터 세트 크기 150


In [34]:
for i in kfold.split(features):
    print(i)

(array([ 30,  31,  32,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,
        43,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,
        56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,
        69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,
        82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,
        95,  96,  97,  98,  99, 100, 101, 102, 103, 104, 105, 106, 107,
       108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120,
       121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133,
       134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146,
       147, 148, 149]), array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]))
(array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  60,  61,  62,  63,

In [23]:
label[[ 30,  31,  32,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,
        43,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,
        56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,
        69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,
        82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,
        95,  96,  97,  98,  99, 100, 101, 102, 103, 104, 105, 106, 107,
       108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120,
       121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133,
       134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146,
       147, 148, 149]]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [48]:
import pandas as pd
iris = load_iris()
iris_df = pd.DataFrame(data=iris.data, columns=iris.features_names)
iris_df['label'] = iris.target
iris_df['label'].value_counts()

AttributeError: features_names

In [45]:
kfold = KFold(n_splits=3)
n_iter = 0
for train_index, test_index in kfold.split(iris_df):
    n_iter += 1
    label_train = iris_df['label'].iloc[train_index]
    label_test = iris_df['label'].iloc[test_index]
    print(f'##교차검증: {n_iter}')
    print(f'학습 레이블 분포: {label_train.value_counts()}')
    print(f'검증 레이블 분포: {label_test.value_counts()}')

NameError: name 'iris_df' is not defined

In [35]:
n_iter = 0
for train_index, test_index in kfold.split(features):
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_tset = label[train_index], label[test_index]
    
    dt_clf.fit(X_train, y_train)
    pred = dt_clf.predict(X_test)
    n_iter += 1
    
    accuracy = np.round(accuracy_score(y_test, pred), 4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n#{0} 교차검증 정확도 : {1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
    print(f'{n_iter} 검증세트 인덱스: {test_index}')
    cv_accuracy.append(accuracy)
    
    print('\n## 평균 검증 정확도:', np.mean(cv_accuracy))


#1 교차검증 정확도 : 1.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
1 검증세트 인덱스: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

## 평균 검증 정확도: 1.0

#2 교차검증 정확도 : 0.6667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
2 검증세트 인덱스: [30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

## 평균 검증 정확도: 0.83335

#3 교차검증 정확도 : 0.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
3 검증세트 인덱스: [60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

## 평균 검증 정확도: 0.5555666666666667

#4 교차검증 정확도 : 0.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
4 검증세트 인덱스: [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

## 평균 검증 정확도: 0.416675

#5 교차검증 정확도 : 0.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
5 검증세트 인덱스: [120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도: 0.33334


#### sklearn.model_selection.StratifiedKFold
* class sklearn.model_selection.StratifiedKFold(n_splits=5, *, shuffle=False, random_state=None)


In [41]:
 from sklearn.model_selection import StratifiedKFold

In [43]:
skf = StratifiedKFold(n_splits=3)
n_iter = 0

for train_index, test_index in skf.split(features, label):
    n_iter += 1
    label_train = iris['label'].iloc[train_index]
    label_test = iris['label'].iloc[test_index]
    print(f'## 교차검증: {n_iter}')
    print(f'학습 레이블 데이터 분포: {label_train.value_counts()}')
    print(f'검증 레이블 데이터 분포: {label_test.value_counts()}')

KeyError: 'label'

In [49]:
from sklearn.model_selection import StratifiedKFold
import pandas as pd

iris_df = pd.DataFrame(features, columns=iris.feature_names)
iris_df['label'] = iris.target

skf = StratifiedKFold(n_splits=3)

cv_accuracy = []

n_iter = 0
for train_index, test_index in skf.split(features , label):
    # print(test_index)
    n_iter +=1
    label_train = iris_df['label'].iloc[train_index]
    label_test = iris_df['label'].iloc[test_index]
    print(f'## 교차검증 {n_iter} ')
    print(label_train.value_counts())

## 교차검증 1 
label
2    34
0    33
1    33
Name: count, dtype: int64
## 교차검증 2 
label
1    34
0    33
2    33
Name: count, dtype: int64
## 교차검증 3 
label
0    34
1    33
2    33
Name: count, dtype: int64


In [50]:
skf = StratifiedKFold(n_splits=3)
cv_accuracy = []

n_iter = 0
for train_index, test_index in skf.split(features, label):
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    
    dt_clf.fit(X_train, y_train)
    pred = dt_clf.predict(X_test)
    n_iter += 1

accuracy = np.round(accuracy_score(y_test, pred), 4)
train_size = X_train.shape[0]
test_size = X_test.shape[0]

print(f'{0} 교차 검증 정확도: {1}, 학습데이터 크기: {2}, 검증 데이터 크기: {3}')
cv_accuracy.append(accuracy)

print(f'평균 검증 정확도: {np.mean(cv_accuracy)}')
    

0 교차 검증 정확도: 1, 학습데이터 크기: 2, 검증 데이터 크기: 3
평균 검증 정확도: 0.98


#### sklearn.model_selection.cross_val_score

* sklearn.model_selection.cross_val_score(estimator=알고리즘명, X, y=None, *, groups=None, scoring=None, cv=None, n_jobs=None, verbose=0, fit_params=None, pre_dispatch='2*n_jobs', error_score=nan)

* estimatorestimator object implementing ‘fit’
The object to use to fit the data.

>>> from sklearn import datasets, linear_model
>>> from sklearn.model_selection import cross_val_score
>>> diabetes = datasets.load_diabetes()
>>> X = diabetes.data[:150]
>>> y = diabetes.target[:150]
>>> lasso = linear_model.Lasso()
>>> print(cross_val_score(lasso, X, y, cv=3))
[0.3315057  0.08022103 0.03531816]
